src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -66,60 +109,244 @@ protected AbstractEntityAIStructure(@NotNull final J job)
         super(job);
         this.registerTargets(
           /**
-           * Check if we have to build something.
-           */
-          new AITarget(this::isThereAStructureToBuild, () -> AIState.START_BUILDING),
+          * Select the appropriate State to do next.
+          */
+          new AITarget(START_BUILDING, this::startBuilding),
           /**
-           * Select the appropriate State to do next.
+           * Check if we have to build something.
            */
-          new AITarget(AIState.START_BUILDING, this::startBuilding),
+          new AITarget(IDLE, this::isThereAStructureToBuild, () -> AIState.START_BUILDING),
           /**
            * Clear out the building area.
-           * todo: implement
            */
-          new AITarget(AIState.CLEAR_STEP, generateSchematicIterator(this::clearStep, AIState.BUILDER_STRUCTURE_STEP)),
+          new AITarget(CLEAR_STEP, generateStructureGenerator(this::clearStep, AIState.BUILDING_STEP)),
           /**
            * Build the structure and foundation of the building.
-           * todo: implement
            */
-          new AITarget(AIState.BUILDING_STEP, () -> AIState.IDLE),
+          new AITarget(BUILDING_STEP, generateStructureGenerator(this::structureStep, AIState.DECORATION_STEP)),
           /**
            * Decorate the AbstractBuilding with torches etc.
-           * todo: implement
            */
-          new AITarget(AIState.DECORATION_STEP, () -> AIState.IDLE),
+          new AITarget(DECORATION_STEP, generateStructureGenerator(this::decorationStep, AIState.COMPLETE_BUILD)),
           /**
            * Spawn entities on the structure.
-           * todo: implement
            */
-          new AITarget(AIState.SPAWN_STEP, () -> AIState.IDLE),
+          new AITarget(SPAWN_STEP, () -> AIState.IDLE),
           /**
            * Finalize the building and give back control to the ai.
-           * todo: implement
            */
-          new AITarget(AIState.COMPLETE_BUILD, () -> AIState.IDLE)
+          new AITarget(COMPLETE_BUILD, this::completeBuild)
         );
     }
 
+    private AIState completeBuild()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "completeBuild" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)